llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp

```diff
@@ -30,7 +63,28 @@ SYCLFixupESIMDKernelWrapperMDPass::run(Module &M, ModuleAnalysisManager &MAM) {
       sycl::utils::traverseCallgraphUp(
           &F,
           [&](Function *GraphNode) {
-            if (!llvm::esimd::isESIMD(*GraphNode)) {
+            if (!llvm::esimd::isESIMD(*GraphNode) &&
+                llvm::esimd::isKernel(*GraphNode)) {
+
+              // Demangle the caller name to check if the function is called
+              // from RoundedRangeKernel.
+              StringRef MangledName = GraphNode->getName();
+              llvm::itanium_demangle::ManglingParser<SimpleAllocator> Parser(
+                  MangledName.begin(), MangledName.end());
+              itanium_demangle::Node *AST = Parser.parse();
+              if (!AST ||
+                  AST->getKind() != itanium_demangle::Node::KSpecialName) {
+                return;
+              }
+
+              itanium_demangle::OutputBuffer NameBuf;
+              AST->print(NameBuf);
+              StringRef Name(NameBuf.getBuffer(), NameBuf.getCurrentPosition());
+
+              if (!Name.contains("sycl::_V1::detail::RoundedRangeKernel<")) {
```

Comment:
@rdeodhar - do you know if there is a simpler/better way to detect if the function is RoundedRangeKernel, without using itanium_demangle? E.g. use some special attribute?

Reply:
There is no attribute that I know of.